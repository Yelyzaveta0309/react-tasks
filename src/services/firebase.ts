import { initializeApp } from "firebase/app";
import { 
    createUserWithEmailAndPassword, 
    getAuth, 
    signInWithEmailAndPassword, 
    signOut } from "firebase/auth";

import {getDatabase, ref} from 'firebase/database';

const firebaseConfig = {
  apiKey: "AIzaSyBaGyOi2iKANasXQXLCGNwiA9ZQduk56eE",
  authDomain: "react-lesson-6703f.firebaseapp.com",
  projectId: "react-lesson-6703f",
  storageBucket: "react-lesson-6703f.appspot.com",
  messagingSenderId: "1089496178670",
  appId: "1:1089496178670:web:918b3d04b39fba05a1b8ff"
};

const firebase = initializeApp(firebaseConfig);

export const auth = getAuth(firebase);

export const signUp = async(email: string, password: string) => 
    await createUserWithEmailAndPassword(auth, email, password);

export const logIn = async(email: string, password: string) => 
    await signInWithEmailAndPassword(auth, email, password);

export const logOut = async ()  => await signOut(auth);

const db = getDatabase(firebase);

export const userRef = ref(db, 'user');
export const chatsRef = ref(db, 'chats');

export const getChatsById = (chatId: string) => ref(db, `chats/${chatId}`);
export const getMessageListById = (chatId: string) => 
    ref(db, `chats/${chatId}/messageList/`);
