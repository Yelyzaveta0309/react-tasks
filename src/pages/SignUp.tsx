import React, { FC, useState } from 'react';
import { signUp } from '../services/firebase';
import { useNavigate } from 'react-router-dom';

export const SignUp: FC = () => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [error, setError] = useState('');

  // const navigate = useNavigate();

  const handleSubmit = async(e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    setError('');
    try{
      await signUp(email, password);
      // navigate("/signin");

    }catch(err){
      setError((err as Error).message)
    }
  };

  return (
    <>
      <h2>Sign Up</h2>
      <hr />
      <form onSubmit={handleSubmit}>
        <label htmlFor="email">Логин:</label>
        <input
          name="email" id="email"
          type="email"
          onChange={(e) => setEmail(e.target.value)}
          value={email}
        />
        <label htmlFor="password">Пароль:</label>
        <input
          name="password" id="password"
          type="password"
          onChange={(e) => setPassword(e.target.value)}
          value={password}
        />
        <br />
        <button>sign up</button>

        {error && <p style={{ color: 'red' }}>{error}</p>}
      </form>
    </>
  );
};