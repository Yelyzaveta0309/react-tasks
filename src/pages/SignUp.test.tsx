import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import { SignUp } from './SignUp';
import '@testing-library/jest-dom';
import { signUp } from '../services/firebase';

jest.mock('../services/firebase', () => ({
  signUp: jest.fn(),
}));

describe('SignUp', () => {

  it('renders the SignUp form', () => {
    const { getByText } = render(<SignUp />);
    expect(getByText('Sign Up')).toBeInTheDocument();
    expect(getByText('Логин:')).toBeInTheDocument();
    expect(getByText('Пароль:')).toBeInTheDocument();
    expect(getByText('sign up')).toBeInTheDocument();
  });

  it('should render correctly', () => {
    const { container } = render(<SignUp />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('calls signUp function with correct arguments when form is submitted', async () => {
    const { getByLabelText, getByText } = render(<SignUp />);
    const emailInput = getByLabelText('Логин:');
    const passwordInput = getByLabelText('Пароль:');
    const submitButton = getByText('sign up');

    fireEvent.change(emailInput, { target: { value: 'test@example.com' } });
    fireEvent.change(passwordInput, { target: { value: 'password123' } });
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(signUp).toHaveBeenCalledWith('test@example.com', 'password123');
    });
  });

  it('should show error message when sign up fails', async () => {
    const errorMessage = 'Sign up failed';
    const mockSignUp = signUp as jest.MockedFunction<typeof signUp>;
    mockSignUp.mockRejectedValueOnce(new Error(errorMessage));
    const { getByLabelText, getByText } = render(<SignUp />);

    const emailInput = getByLabelText('Логин:');
    const passwordInput = getByLabelText('Пароль:');
    const signUpButton = getByText('sign up');

    fireEvent.change(emailInput, { target: { value: 'test@example.com' } });
    fireEvent.change(passwordInput, { target: { value: 'test1234' } });
    fireEvent.click(signUpButton);

    await waitFor(() => {
      expect(mockSignUp).toHaveBeenCalledWith('test@example.com', 'test1234');
      expect(getByText(errorMessage)).toBeInTheDocument();
    });
  });

});
