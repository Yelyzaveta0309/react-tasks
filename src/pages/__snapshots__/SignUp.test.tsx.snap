// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp component should render correctly 1`] = `
<h2>
  Sign Up
</h2>
`;

exports[`SignUp should render correctly 1`] = `
<h2>
  Sign Up
</h2>
`;
